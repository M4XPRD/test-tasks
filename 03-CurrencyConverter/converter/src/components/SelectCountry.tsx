/* eslint-disable react/jsx-props-no-spreading */
import '../navigator.d';
import {
  Autocomplete, Grid, TextField,
} from '@mui/material';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../slices';
import { SetExchangeType } from '../slices/currenciesSlice';
import { Country } from '../slices/countriesSlice';

interface Props {
  currencyValue: string;
  label: string;
  setExchange: SetExchangeType;
  exchangeOption?: string;
}

interface SetExchangeAction {
  type: string;
  payload: {
    newValue: string;
    exchangeOption?: string;
  };
}

const SelectCountry = (props: Props) => {
  const {
    currencyValue, label, setExchange, exchangeOption,
  } = props;
  const dispatch = useDispatch();
  const countriesList = useSelector((state: RootState) => state.countries.countriesList);

  const formOptions = countriesList.map((country: object) => {
    const countryTypes = country as Country;
    const os = navigator.userAgentData.platform;
    const [currency] = Object.keys(countryTypes.currencies);
    const countryName = countryTypes.name.common;
    const currencyData = `${os === 'Windows' ? '' : countryTypes.flag} ${currency} â€” ${countryName}`;
    return currencyData.trim();
  });

  return (
    <Grid item xs={12} md={3.5}>
      <Autocomplete
        value={currencyValue}
        isOptionEqualToValue={(option, value) => option.trim() === value.trim()}
        disableClearable
        onChange={(e, newValue) => {
          const action: SetExchangeAction = setExchange({ newValue, exchangeOption });
          dispatch(action);
        }}
        options={formOptions}
        renderInput={(params) => (
          <TextField
            {...params}
            label={label}
            type="text"
            autoComplete="off"
          />
        )}
      />
    </Grid>
  );
};

SelectCountry.defaultProps = {
  exchangeOption: undefined,
};

export default SelectCountry;
