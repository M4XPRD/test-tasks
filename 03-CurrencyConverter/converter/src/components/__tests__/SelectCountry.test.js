import {
  render, screen, waitFor,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import userEvent from '@testing-library/user-event';
import SelectCountry from '../SelectCountry';
import store from '../../slices/index';

describe('SelectCountry component', () => {
  it('Should update the selected country in Autocomplete', async () => {
    const setExchangeMock = jest.fn();
    const label = 'Select a country';
    const currencyValue = 'USD';
    const exchangeOption = 'from';

    render(
      <Provider store={store}>
        <SelectCountry
          label={label}
          currencyValue={currencyValue}
          exchangeOption={exchangeOption}
          setExchange={setExchangeMock}
        />
      </Provider>,
    );

    const input = screen.getByRole('combobox');
    userEvent.clear(input);
    userEvent.type(input, 'AUD');
    userEvent.keyboard('[ArrowDown]');
    userEvent.keyboard('[Enter]');
    await waitFor(() => expect(input).toHaveValue('AUD'));
  });
});
