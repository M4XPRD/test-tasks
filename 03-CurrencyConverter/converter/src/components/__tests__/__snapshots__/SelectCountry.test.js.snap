// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectCountry component SelectCountry snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3.5 css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1h51icj-MuiAutocomplete-root"
          data-testid="autocomplete"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            inputmode="none"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                role="combobox"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3.5 css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1h51icj-MuiAutocomplete-root"
        data-testid="autocomplete"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          inputmode="none"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
